pageserver/src/tenant/storage_layer/inmemory_layer.rs

```diff
@@ -54,6 +54,12 @@ pub struct InMemoryLayer {
     /// Writes are only allowed when this is `None`.
     end_lsn: OnceLock<Lsn>,
 
+    /// Used for traversal path. Cached representation of the in-memory layer before frozen.
+    local_path_str: Arc<str>,
+
+    /// Used for traversal path. Cached representation of the in-memory layer after frozen.
+    frozen_local_path_str: OnceLock<Arc<str>>,
```

Comment:
I don't really see why caching this is worth the extra complexity, but I don't feel strongly about it.

Reply:
Otherwise we clone a String every time we do get page at lsn, so I would prefer doing a cache here so that we have `Arc<str>` created only once for each layer (or twice for in-mem layer)

Reply:
As an alternative, would it be possible to `cfg!(feature = "testing")` all of this path stuff? I think it's only printed on failures when the testing feature is enabled.

Reply:
I feel it is still important to get a good backtrace of all information we have if there is a reconstruct page error in prod. Therefore, I would prefer caching it there :) Or we can use a different identifier for the in-memory layer (i.e., do not include end LSN).